### PR TITLE
EventEngine::RunAt(): Weighted Target LB Policy removal timer

### PR DESCRIPTION
**Pull request by drfloob**
There may be a pre-existing race here if the timer callback is fired while the DelayedRemovalTimer is being orphaned. A timer mutex would solve it, but I'm not sure it's necessary.

Potential sequence of events:
T1: Orphan: and `timer_handle_.has_value() = true`, if block entered.
T2: OnTimerLocked: `timer_handle.reset()` succeeds
T1: Orphan continued: attempts to call `GetDefaultEventEngine()->Cancel(*timer_handle_);`

### CODE REVIEW

File: src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc [hunk 1]
```diff
@@ -444,44 +445,32 @@ void WeightedTargetLb::UpdateStateLocked() {
 WeightedTargetLb::WeightedChild::DelayedRemovalTimer::DelayedRemovalTimer(
     RefCountedPtr<WeightedTargetLb::WeightedChild> weighted_child)
     : weighted_child_(std::move(weighted_child)) {
-  GRPC_CLOSURE_INIT(&on_timer_, OnTimer, this, nullptr);
-  Ref().release();
-  grpc_timer_init(&timer_, ExecCtx::Get()->Now() + kChildRetentionInterval,
-                  &on_timer_);
+  timer_handle_ = GetDefaultEventEngine()->RunAt(
+      absl::Now() + kChildRetentionInterval, [self = Ref()] {
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc:449]: Doesn't this lambda need to be `mutable` in order to do the `std::move()` into the nested lambda below?

**Review by markdroth — COMMENTED** [review]
I don't follow your comment about the pre-existing race.  In what sequence of events would this race occur?

File: src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc (hunk 1 above)
Comment (drfloob) [reply src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc:449]: Yes, you're right. I've replaced it with `self = Ref()`. Even though the function will only be called once, making the arg mutable feels like the wrong choice.

**Comment by drfloob** [comment]
I've updated the description to explain the race. Thanks for the review.

File: src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc (hunk 1 above)
Comment (drfloob) [reply src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc]: Done. As discussed, the fact that [this behavior](https://godbolt.org/z/5sn1dzhcE) sometimes surprises folks is why I prefer to not use mutable, but I'm fine to use here it and move on given that it's not a problem in this case.

**Review by markdroth — APPROVED** [review]

**Pull request merged**